docs/canonical_testing.md

```diff
@@ -0,0 +1,186 @@
+# Canonical testing overview
+
+We want to make it trivial for everyone to discover how to run tests, and to run
+tests the same way.
+
+In order to do this we will provide:
+
+* A new standardized command to invoke any test job, which calls
+  - `k8stest`
+  - This knows how to map jobs into a job class.
+  - This knows how to schedule a job locally, on docker or kubernetes.
+* Standardize the command for each job class
+  - `hack/e2e.go`, `hack/test-go.sh`, etc.
+  - This command encapsulates everything needed to run a job, potentially with
+    some customization passed in as an argument.
+* Create a standardized image for each job class.
+  - `kubernetes-e2e`, `test-go`, etc
+  - This image encapsulates all the dependencies required to run the job at any
+    given state of the repo and allows for the required customization.
+* A standardized yaml files for each job that customizes the class.
+  - `kubernetes-e2e-gce.yaml`, `kubernetes-gke-slow`, etc.
+  - This defines how each job customizes the class. For example we run the same
+    unit tests at different apis, and we run the same e2e flow with differnt
+    ginkgo foci.
+
+This will allow:
+
+* There is a deterministic way to map a job to a class to commands in this
+  class.
+* Different people can confidently run the job the same way jenkins will do so.
+* All jobs can run inside or outside a container.
+* All jobs can run on or off a cluster.
+
+
+## k8stest
+
+The top level command allows
+
+
+* A {job: (class, configuration)} mapping.
+* A {set: [job]} mapping.
+* Logic to map configuration to each class.
+* Flags to pass in options
+  - where and whether to upload results.
+  - any required configuration to access a cluster and/or provider
+
+Some basic examples:
+
+```
+k8stest test-go  # Run the unit tests
+k8stest test-integration  # Run the integration tests
+k8stest verify  # Run the verify tests
+k8stest e2e-gce  # Run the e2e tests on gce
+k8stest e2e-gke-serial  # Run the serial tests on gke
+```
+
+Customizing how a specific job runs:
+```
+k8stest test-go -- pkg/api pkg/kubelet # Run only the api and kubelet unit tests
+k8stest test-integration -- Deadline  # Run only Deadline integration tests
+k8stest verify -- godeps  # Run only the godep verification
+k8stest e2e-gce -- --push --test  # Push to an existing cluster and run tests
+k8stest e2e-gke-serial -- --skip=Daemon  # Skip the Daemon tests
```

Comment:
True, but this is largely an artifact of using ginkgo in one but not the other location.